src/Processors/Sources/SourceFromChunks.cpp

```diff
@@ -3,20 +3,61 @@
 namespace DB
 {
 
-SourceFromChunks::SourceFromChunks(Block header, Chunks chunks_)
-    : SourceFromChunks(header, std::make_shared<Chunks>(std::move(chunks_)), true)
-{}
-
-SourceFromChunks::SourceFromChunks(Block header, std::shared_ptr<Chunks> chunks_)
-    : SourceFromChunks(header, chunks_, false)
-{}
-
-SourceFromChunks::SourceFromChunks(Block header, std::shared_ptr<Chunks> chunks_, bool move_from_chunks_)
-    : ISource(std::move(header))
-    , chunks(chunks_)
-    , it(chunks->begin())
-    , move_from_chunks(move_from_chunks_)
+SourceFromChunks::SourceFromChunks(Block header, Chunks && chunks_, std::optional<Chunk> && chunk_totals_, std::optional<Chunk> && chunk_extremes_)
+    : ISource(header)
+    , chunks(std::move(chunks_))
+    , it(chunks.begin())
 {
+    if (chunk_totals_ != std::nullopt)
+    {
+        outputs.emplace_back(header, this);
+        chunk_totals = std::move(chunk_totals_);
+        output_totals = &outputs.back();
+    }
+
+    if (chunk_extremes_ != std::nullopt)
+    {
+        outputs.emplace_back(header, this);
+        chunk_extremes = std::move(chunk_extremes_);
+        output_extremes = &outputs.back();
+    }
+}
+
+SourceFromChunks::Status SourceFromChunks::prepare()
+{
+    if (!finished_chunks)
+    {
+        Status status = ISource::prepare();
+
+        if (status != Status::Finished)
+            return status;
+
+        finished_chunks = true;
+    }
+
+    if (getTotalsPort())
+    {
+        /// This logic force-pushes the data into the port instead of checking the port status first (isFinished(), canPush()). Seems to
```

Comment:
it seems easy to fix, let's do it straight away

Reply:
I simplified further, that logic is gone now.